File: Algo/Leetcode/069Sqrt.cpp
// binary search
class Solution {
public:
    int mySqrt(int x) {
        long l = 1; long r = x;  // in case overflow
        while(l<=r){
            long long m = (l+r)/2;
            long long y = m *m;
            if(y == x) return m;
            else if(y>x) r = m-1;
            else l = m+1;
        }
        return r;
    }
};

// NewTon
class Solution {
public:
    int mySqrt(int x) {
        double x0 = (double)x;
        double y = 0.5 * ((double)x + 1.0);
        while(abs(y-x0)>1e-7){
            x0 = y;
            y = 0.5 * (y + (double)x/y);
        }
        return (int)y;
    }
};

File: Algo/Leetcode/069Sqrt.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        l = 0
        r = x
        while l <= r:
            mid = (l+r) / 2
            if mid**2 <= x and (mid+1)**2 > x:
                return mid
            elif mid ** 2 > x:
                r = mid-1
            else:
                l = mid+1

        return r
